openhtf/plugs/__init__.py

```diff
@@ -196,52 +197,70 @@ def tearDown(self):
     pass
 
 
-class RemotePlug(xmlrpcutil.TimeoutProxyMixin, xmlrpcutil.BaseServerProxy,
-                 sockjs.tornado.SockJSConnection):
+class RemotePlug(xmlrpcutil.TimeoutProxyServer):
   """Remote interface to a Plug.
 
-  This class is used by the frontend server to poke at Plugs that are currently
-  in use by a remotely running OpenHTF Test.  It also provides a SockJS
-  interface as a SockJSConnection for frontend clients to interact with.
+  This class is used by the web GUI to call methods on plugs that are currently
+  in use by a remotely running OpenHTF test.
   """
 
-  def __init__(self, host, port, plug_name, session):
-    super(RemotePlug, self).__init__((host, port))
-    sockjs.tornado.SockJSConnection.__init__(self, session)
+  class RemotePlugHandler(tornado.web.RequestHandler):
+    """Handles HTTP POST requests directed at remote plugs."""
+
+    def initialize(self, remote_plug):
+      self.remote_plug = remote_plug
+
+    def post(self):
+      """Called when we receive a JSON message from a frontend client."""
+      try:
+        data = json.loads(self.request.body)
+        method = data['method']
+        args = data['args']
+      except (KeyError, ValueError):
+        self._write_error('Malformed JSON request')
+        return
+
+      try:
+        result = getattr(self.remote_plug, method)(*args)
+      except (AttributeError, TypeError, ValueError):
+        self._write_error('Error in RPC request')
+        return
+
+      self.write(json.dumps(result))
+
+    def _write_error(self, message):
+      """Helper that returns an HTTP 500 error and logs the error details."""
+      _LOG.warning('%s: %s', message, self.request.body, exc_info=True)
+      self.set_status(500)
+      self.write('RemotePlug error: %s.' % message)
+
+  def __init__(self, host, port, plug_name):
+    super(RemotePlug, self).__init__('http://%s:%s' % (host, port))
+    self.plug_name = plug_name
+
     # Grab exactly the attrs we want, the server has all plugs', not just ours.
-    for method in self.system.listMethods():
+    for method in self.list_methods():
       remote_attr = RemoteAttribute.from_method_string(method)
       if remote_attr.class_name == plug_name:
         if hasattr(type(self), remote_attr.name):
-          _LOG.warning('Skipping predefined attribute "%s" for remote access.',
+          _LOG.warning('Skipping preexisting remote plug attribute "%s".',
                        remote_attr.name)
         else:
-          setattr(self, remote_attr.name,
-                  functools.partial(self.__request, method))
+          setattr(self, remote_attr.name, getattr(self, method))
+
+  def list_methods(self):
```

Comment:
Why is this function here? `self.system` should work since `__getattr__` won't get invoked for attributes that exist so this should be the same as just doing `self.system.listMethods()` in other functions directly